### ISSUE TITLE
Problem with Google Opauth redirect after app authorization

### ISSUE THREAD

**Issue by skyserpent** [uzyn/cakephp-opauth#28]
My demo Opauth CakePHP install is sitting at 'localhost/cake' and I've also added a 2.x Cake library to the app folder that comes with the demo. I added my client ID and secret and got the app to authorize with Google, but here is where I'm getting redirected to after that:

http://localhost/cake/auth/google/oauth2callback?code=#xxxxxxx

and here is the content being displayed in the body:

Notice (8): Undefined property: stdClass::$picture [APP\Plugin\Opauth\Strategy\Google\GoogleStrategy.php, line 89]
Code Context
GoogleStrategy::oauth2callback() - APP\Plugin\Opauth\Strategy\Google\GoogleStrategy.php, line 89
OpauthStrategy::callAction() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\OpauthStrategy.php, line 188
Opauth::run() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\Opauth.php, line 117
OpauthAppController::index() - APP\Plugin\Opauth\Controller\OpauthAppController.php, line 31
ReflectionMethod::invokeArgs() - [internal], line ??
Controller::invokeAction() - CORE\Cake\Controller\Controller.php, line 490
Dispatcher::_invoke() - CORE\Cake\Routing\Dispatcher.php, line 187
Dispatcher::dispatch() - CORE\Cake\Routing\Dispatcher.php, line 162
[main] - APP\webroot\index.php, line 92

Warning (2): session_start() [function.session-start]: Cannot send session cache limiter - headers already sent (output started at C:\xampp\htdocs\cake\lib\Cake\Utility\Debugger.php:804) [APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\OpauthStrategy.php, line 174]
Code Context
session_start - [internal], line ??
OpauthStrategy::shipToCallback() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\OpauthStrategy.php, line 174
OpauthStrategy::callback() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\OpauthStrategy.php, line 119
GoogleStrategy::oauth2callback() - APP\Plugin\Opauth\Strategy\Google\GoogleStrategy.php, line 100
OpauthStrategy::callAction() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\OpauthStrategy.php, line 188
Opauth::run() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\Opauth.php, line 117
OpauthAppController::index() - APP\Plugin\Opauth\Controller\OpauthAppController.php, line 31
ReflectionMethod::invokeArgs() - [internal], line ??
Controller::invokeAction() - CORE\Cake\Controller\Controller.php, line 490
Dispatcher::_invoke() - CORE\Cake\Routing\Dispatcher.php, line 187
Dispatcher::dispatch() - CORE\Cake\Routing\Dispatcher.php, line 162
[main] - APP\webroot\index.php, line 92

Warning (2): Cannot modify header information - headers already sent by (output started at C:\xampp\htdocs\cake\lib\Cake\Utility\Debugger.php:804) [APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\OpauthStrategy.php, line 306]
Code Context
header - [internal], line ??
OpauthStrategy::redirect() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\OpauthStrategy.php, line 306
OpauthStrategy::shipToCallback() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\OpauthStrategy.php, line 177
OpauthStrategy::callback() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\OpauthStrategy.php, line 119
GoogleStrategy::oauth2callback() - APP\Plugin\Opauth\Strategy\Google\GoogleStrategy.php, line 100
OpauthStrategy::callAction() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\OpauthStrategy.php, line 188
Opauth::run() - APP\Plugin\Opauth\Vendor\Opauth\lib\Opauth\Opauth.php, line 117
OpauthAppController::index() - APP\Plugin\Opauth\Controller\OpauthAppController.php, line 31
ReflectionMethod::invokeArgs() - [internal], line ??
Controller::invokeAction() - CORE\Cake\Controller\Controller.php, line 490
Dispatcher::_invoke() - CORE\Cake\Routing\Dispatcher.php, line 187
Dispatcher::dispatch() - CORE\Cake\Routing\Dispatcher.php, line 162
[main] - APP\webroot\index.php, line 92

**Comment by skyserpent** [comment]
MY BOOTSTRAP FILE:

<?php
/**
- This file is loaded automatically by the app/webroot/index.php file after core.php
  *
- This file should load/create any application wide configuration settings, such as 
- Caching, Logging, loading additional configuration files.
  *
- You should also use this file to include any files that provide global functions/constants
- that your application uses.
  *
- PHP 5
  *
- CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
- Copyright 2005-2012, Cake Software Foundation, Inc. (http://cakefoundation.org)
  *
- Licensed under The MIT License
- Redistributions of files must retain the above copyright notice.
  *
- @copyright     Copyright 2005-2012, Cake Software Foundation, Inc. (http://cakefoundation.org)
- @link          http://cakephp.org CakePHP(tm) Project
- @package       app.Config
- @since         CakePHP(tm) v 0.10.8.2117
- @license       MIT License (http://www.opensource.org/licenses/mit-license.php)
  */

/**
- Cache Engine Configuration
- Default settings provided below
  *
- File storage engine.
  *
-   Cache::config('default', array(
-      'engine' => 'File', //[required]
-      'duration'=> 3600, //[optional]
-      'probability'=> 100, //[optional]
-      'path' => CACHE, //[optional] use system tmp directory - remember to use absolute path
-      'prefix' => 'cake_', //[optional]  prefix every cache file with this string
-      'lock' => false, //[optional]  use file locking
-      'serialize' => true, // [optional]
-      'mask' => 0666, // [optional] permission mask to use when creating cache files
-  ));
  *
- APC (http://pecl.php.net/package/APC)
  *
-   Cache::config('default', array(
-      'engine' => 'Apc', //[required]
-      'duration'=> 3600, //[optional]
-      'probability'=> 100, //[optional]
-      'prefix' => Inflector::slug(APP_DIR) . '_', //[optional]  prefix every cache file with this string
-  ));
  *
- Xcache (http://xcache.lighttpd.net/)
  *
-   Cache::config('default', array(
-      'engine' => 'Xcache', //[required]
-      'duration'=> 3600, //[optional]
-      'probability'=> 100, //[optional]
-      'prefix' => Inflector::slug(APP_DIR) . '_', //[optional] prefix every cache file with this string
-      'user' => 'user', //user from xcache.admin.user settings
-      'password' => 'password', //plaintext password (xcache.admin.pass)
-  ));
  *
- Memcache (http://memcached.org/)
  *
-   Cache::config('default', array(
-      'engine' => 'Memcache', //[required]
-      'duration'=> 3600, //[optional]
-      'probability'=> 100, //[optional]
-      'prefix' => Inflector::slug(APP_DIR) . '_', //[optional]  prefix every cache file with this string
-      'servers' => array(
-          '127.0.0.1:11211' // localhost, default port 11211
-      ), //[optional]
-      'persistent' => true, // [optional] set this to false for non-persistent connections
-      'compress' => false, // [optional] compress data in Memcache (slower, but uses less memory)
-  ));
  *
-  Wincache (http://php.net/wincache)
  *
-   Cache::config('default', array(
-      'engine' => 'Wincache', //[required]
-      'duration'=> 3600, //[optional]
-      'probability'=> 100, //[optional]
-      'prefix' => Inflector::slug(APP_DIR) . '_', //[optional]  prefix every cache file with this string
-  ));
  */
  Cache::config('default', array('engine' => 'File'));

/**
- The settings below can be used to set additional paths to models, views and controllers.
  *
- App::build(array(
-     'Model'                     => array('/path/to/models', '/next/path/to/models'),
-     'Model/Behavior'            => array('/path/to/behaviors', '/next/path/to/behaviors'),
-     'Model/Datasource'          => array('/path/to/datasources', '/next/path/to/datasources'),
-     'Model/Datasource/Database' => array('/path/to/databases', '/next/path/to/database'),
-     'Model/Datasource/Session'  => array('/path/to/sessions', '/next/path/to/sessions'),
-     'Controller'                => array('/path/to/controllers', '/next/path/to/controllers'),
-     'Controller/Component'      => array('/path/to/components', '/next/path/to/components'),
-     'Controller/Component/Auth' => array('/path/to/auths', '/next/path/to/auths'),
-     'Controller/Component/Acl'  => array('/path/to/acls', '/next/path/to/acls'),
-     'View'                      => array('/path/to/views', '/next/path/to/views'),
-     'View/Helper'               => array('/path/to/helpers', '/next/path/to/helpers'),
-     'Console'                   => array('/path/to/consoles', '/next/path/to/consoles'),
-     'Console/Command'           => array('/path/to/commands', '/next/path/to/commands'),
-     'Console/Command/Task'      => array('/path/to/tasks', '/next/path/to/tasks'),
-     'Lib'                       => array('/path/to/libs', '/next/path/to/libs'),
-     'Locale'                    => array('/path/to/locales', '/next/path/to/locales'),
-     'Vendor'                    => array('/path/to/vendors', '/next/path/to/vendors'),
-     'Plugin'                    => array('/path/to/plugins', '/next/path/to/plugins'),
- ));
  *
  */

/**
- Custom Inflector rules, can be set to correctly pluralize or singularize table, model, controller names or whatever other
- string is passed to the inflection functions
  *
- Inflector::rules('singular', array('rules' => array(), 'irregular' => array(), 'uninflected' => array()));
- Inflector::rules('plural', array('rules' => array(), 'irregular' => array(), 'uninflected' => array()));
  *
  */

/**
- Plugins need to be loaded manually, you can either load them one by one or all of them in a single call
- Uncomment one of the lines below, as you need. make sure you read the documentation on CakePlugin to use more
- advanced ways of loading plugins
  *
- CakePlugin::loadAll(); // Loads all plugins at once
- CakePlugin::load('DebugKit'); //Loads a single plugin named DebugKit
  *
  */

// Loads Opauth Plugin
CakePlugin::load('Opauth', array('routes' => true, 'bootstrap' => true));

/**
- If your CakePHP app does not reside at DocumentRoot (eg. http://localhost),
- but at a directory below DocumentRoot (eg. http://localhost/your-cake-app)
- Uncomment the next line and replace 'your-cake-app' with your actual path
  */
    Configure::write('Opauth.path', '/cake/auth/');

/**
- Opauth strategy configurations
  */
    // Facebook strategy
    Configure::write('Opauth.Strategy.Facebook', array(
        'app_id' => 'YOUR FACEBOOK APP ID',
        'app_secret' => 'YOUR FACEBOOK APP SECRET'
    ));
  
    // Google strategy
    Configure::write('Opauth.Strategy.Google', array(
        'client_id' => 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
        'client_secret' => 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
    ));
  
    // Twitter strategy
    Configure::write('Opauth.Strategy.Twitter', array(
        'key' => 'YOUR CONSUMER KEY',
        'secret' => 'YOUR CONSUMER SECRET'
    ));
  
    // OpenID strategy
    Configure::write('Opauth.Strategy.OpenID', array(
    ));

**Comment by skyserpent** [comment]
MY ROUTES FILE:

<?php
/**
- Routes configuration
  *
- In this file, you set up routes to your controllers and their actions.
- Routes are very important mechanism that allows you to freely connect
- different urls to chosen controllers and their actions (functions).
  *
- PHP 5
  *
- CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
- Copyright 2005-2012, Cake Software Foundation, Inc. (http://cakefoundation.org)
  *
- Licensed under The MIT License
- Redistributions of files must retain the above copyright notice.
  *
- @copyright     Copyright 2005-2012, Cake Software Foundation, Inc. (http://cakefoundation.org)
- @link          http://cakephp.org CakePHP(tm) Project
- @package       app.Config
- @since         CakePHP(tm) v 0.2.9
- @license       MIT License (http://www.opensource.org/licenses/mit-license.php)
  _/
  /_*
- Here, we are connecting '/' (base path) to controller called 'Pages',
- its action called 'display', and we pass a param to select the view file
- to use (in this case, /app/View/Pages/home.ctp)...
  _/
    Router::connect('/', array('controller' => 'pages', 'action' => 'display', 'home'));
  /_*
- ...and connect the rest of 'Pages' controller's urls.
  _/
    Router::connect('/pages/_', array('controller' => 'pages', 'action' => 'display'));

/**
- Opauth callback
  _/
    Router::connect(
        '/opauth-complete/_', 
        array('controller' => 'users', 'action' => 'opauth_complete')
    );

/**
- Load all plugin routes.  See the CakePlugin documentation on 
- how to customize the loading of plugin routes.
  */
    CakePlugin::routes();

/**
- Load the CakePHP default routes. Remove this if you do not want to use
- the built-in default routes.
  */
    require CAKE . 'Config' . DS . 'routes.php';

**Comment by uzyn** [comment]
@qsoomro You may want to update your Google strategy to the latest from https://github.com/opauth/google/

**Comment by skyserpent** [comment]
Wow quick reply. Thanks. I updated the strategy and now its redirecting me to 'localhost/auth/callback' not to sure why it's doing that. I didn't change any of the files in the meantime.

**Comment by uzyn** [comment]
Make sure you Opauth.path is still set to /cake/auth.

**Comment by skyserpent** [comment]
Hey I solved it buddy. Thanks for your help though you're great. 

Here's the fix:

In the _loadOpauth function in file: OpauthAppController, I added this line at the top:

if (Configure::read('Opauth.callback_url') == '/auth/callback') {
       Configure::write('Opauth.callback_url', Configure::read('Opauth.path').'callback');
}

I would say from this whole experience that updating the sample CakePHP Opauth demo app would be a great idea. Thanks again bud.
